src/policy/marksweepspace/native_ms/global.rs

```diff
@@ -419,6 +431,42 @@ impl<VM: VMBinding> MarkSweepSpace<VM> {
     pub fn get_abandoned_block_lists_in_gc(&self) -> &Mutex<AbandonedBlockLists> {
         &self.abandoned_in_gc
     }
+
+    pub fn release_packet_done(&self) {
+        let old = self.pending_release_packets.fetch_sub(1, Ordering::SeqCst);
+        if old == 1 {
+            let work_packets = self.generate_sweep_tasks();
+            self.scheduler.work_buckets[WorkBucketStage::Release].bulk_add(work_packets);
+        }
+    }
+
+    fn generate_sweep_tasks(&self) -> Vec<Box<dyn GCWork<VM>>> {
+        let space = unsafe { &*(self as *const Self) };
+        let epilogue = Arc::new(RecategorizeBlocks {
```

Comment:
I have a PR here that extracts the epilogue mechanism outside of the space code: https://github.com/mmtk/mmtk-core/pull/877. Do you want me to tidy it up and get it merged so you can base this PR on it?

Reply:
I tried tidying it up locally, but I feel that #877 may not be general enough.  One example is the `pending_release_packets` field I put into `MarkSweepSpace`.  Due to our current mechanism of creating the `ReleaseMutator` work packets, we can't wrap `ReleaseMutator` into a `WorkPacketWithEpilogue`.  Therefore, I just use the space instance as the shared data.  Another thing I worry about is `WorkPacketEpilogue` wraps a `Box<dyn Fn>` inside it (it's better to use `FnOnce` instead).  We may want to mutate the epilogue in some ways after it is created, according to the work packets that will count down the counter.

I also think the "epilogue" may have some capability of collecting data in addition to counting.  For example, each unit of work could "abandon" some blocks to a pool which also plays the role of "epilogue".  Maybe the most general mechanism is a simple wrapper of `AtomicUsize` that provides a few intuitively named methods, and implement something like a countdown latch.

BTW, I have a utility class [`AfterAll`](https://github.com/mmtk/mmtk-ruby/blob/cf70302be2008626fb120e5353880e073b33ff09/mmtk/src/utils.rs#L60) in the Ruby binding.  It is more special purposed, but it's not that hard to implement or use.